content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,64 @@
+---
+title: "A confusing build failure"
+date: 2019-11-07T01:00:09+02:00
+draft: true
+tags: ["testing", "continuous integration"]
+---
+
+{{< figure src="/images/posts/broken_builds/judge-dredd-under-arrest.jpg" title="" >}}
+
+Nowadays it has become quite a common practice to build and check software on machines other than local workstations, most of the time called CI servers(CI stands for [Continuous Integration](https://martinfowler.com/articles/continuousIntegration.html)). This can be done manually or automatically, usually triggered by a pull/merge request or a direct push on the main trunk(in git terminology most of the times named *master*).
+
+The point is that there's a single point where binaries are packaged and code is checked for quality and correctness, so we can merge changes safely and with confidence. 
+
+When a failure appears in a build it should be investigated why did it happen. Did a commit actually break functionality/introduce a bug? Was static code analysis checks too harsh and maybe they must be relaxed a little bit? One of the failures that left me most perplexed was a build that was failing on the CI server unit tests but passed locally! How did this happen? Investigating on this confusing build failure unearthed bigger problems than just a breaking unit
+test.
+
+## Mockito cannot mock final classes
+
+So why was the build failing? When the unit tests ran some of them would fail with an error stating that Mockito(more specifically [mockito-kotlin](https://github.com/nhaarman/mockito-kotlin)) - a library commonly used in Java and Android to provide mock test doubles and for behavior verification - cannot mock final classes.
+
+In order to understand the error let's try to understand first how Mockito works. It creates at runtime objects that extend interfaces or a classes which can be used to verify that some methods were called with specific parameters. If a class that is attempted to be mocked is declared as `final` then the error above will be thrown, as a final class cannot be extended at compile or run time. In a unit test a Kotlin class was mocked and as Kotlin classes are by default
+`final` unless explicitly declared `open`, the error above was thrown.
+
+The culprit was found but we still have not figured out why the tests pass locally but fail on the CI. 
+
+## Maybe Mockito can mock final classes after all
+
+If Mockito cannot mock final classes then why do local tests pass? Well sure there must be some way to mock a final class, and that's exactly what happened by adding in the gradle build file the `org.mockito:mockito-inline` dependency. 
+
+The problem is that the failing module did not compile with the dependency above. Another submodule did compile it, but it was declared on its `build.gradle` as `testImplementation` not `testApi`, meaning that final classes could be mocked on that submodule only, not on the failing module.
```

Comment:
```suggestion
The problem with our case was that the failing module on the CI server did not compile with the above dependency while a submodule on the local workstation did. This unexpected behavior was traced back to the `build.gradle` script of the local workstation where the mockito-inline dependency was set to `testImplementation` and not `testApi`, meaning that final classes could be mocked only on that submodule and not on the failing module.
```

Reply:
@ktsiligkos Actually the behavior is well expected.  A dependency  is declared in gradle as `api` seldom because that way it's transitive something that generally is avoided. 

What I try to highlight here is that mockito-inline was expected to not compile in the top module, but it did compile nevertheless.